akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala

```diff
@@ -0,0 +1,291 @@
+        */
+      override def preStart(): Unit = {
+        printy("preStart")
+        doPull(requestIn)
+        printStatus()
+      }
+
+      /**
+        * This is to tell to requestIn that demand exists on requestOut
+        */
+      setHandler(requestOut, new OutHandler {
+        /**
+          * Called when the output port has received a pull, and therefore ready to emit an element,
+          * i.e. [[GraphStageLogic.push()]] is now allowed to be called on this port.
+          */
+        override def onPull(): Unit = {
+          printy("requestOut: onPull")
+          printStatus()
+        }
+
+        /**
+          * Called when the output port will no longer accept any new elements. After this callback no other callbacks
+          * will be called for this port.
+          */
+        override def onDownstreamFinish(): Unit = {
+          printy("requestOut: onDownstreamFinish")
+          cancel(requestIn)
+        }
+
+      })
+
+      setHandler(requestIn, new InHandler {
+        /**
+          * Called when the input port has a new element available. The actual element can be retrieved via the
+          * [[GraphStageLogic.grab()]] method.
+          */
+        override def onPush(): Unit = {
+          printy("requestIn: onPush")
+          val r = grab(requestIn)
+          emit(requestOut, r, () => doPull(responseIn))
+          inFlight = inFlight :+ r
+          loopDetector += (r -> List(r.uri))
+          printStatus()
+        }
+
+        /**
+          * Called when the input port is finished. After this callback no other callbacks will be called for this port.
+          */
+        override def onUpstreamFinish(): Unit = {
+          printy("requestIn: onUpstreamFinish")
+          complete(requestOut)
+        }
+
+        /**
+          * Called when the input port has failed. After this callback no other callbacks will be called for this port.
+          */
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          printy("requestIn: onUpstreamFailure")
+          fail(requestOut, ex)
+        }
+      })
+
+      /**
+        * Response is available. Check if it is redirect and act accordingly.
+        */
+      setHandler(responseIn, new InHandler {
+        /**
+          * Called when the input port has a new element available. The actual element can be retrieved via the
+          * [[GraphStageLogic.grab()]] method.
+          */
+        override def onPush(): Unit = {
+          printy("responseIn: onPush")
+          val response = grab(responseIn)
+          val maybeRedirect = redirectMapper(settings, inFlight.head, response)
+          if (maybeRedirect.isDefined) {
+            val redirect = maybeRedirect.get
+            printy(loopDetector.toString())
+            val loops = loopDetector(inFlight.head)
+            if (loops.contains(redirect.uri)) {
+              throw InfiniteRedirectLoopException
```

Comment:
I forgot which is better - throw or fail the stage? 
The answer depends on if we want to log this failure... I guess we do.
// cc @drewhk